sdk/schemaregistry/azure-schemaregistry-avroserializer/azure/schemaregistry/serializer/avroserializer/_schema_registry_avro_serializer.py

```diff
@@ -135,12 +147,37 @@ def serialize(self, value, **kwargs):
         try:
             cached_schema = self._user_input_schema_cache[raw_input_schema]
         except KeyError:
-            parsed_schema = avro.schema.parse(raw_input_schema)
+            try:
+                parsed_schema = avro.schema.parse(raw_input_schema)
+            except (
+                SchemaParseException,
+                AvroTypeException,
+                AvroException,
+                InvalidName,
+            ) as e:
+                raise AzureSchemaParseException(e)
             self._user_input_schema_cache[raw_input_schema] = parsed_schema
             cached_schema = parsed_schema
 
+        # TODO: check if there's a more pythonic way, if we need it at all
+        if cached_schema.type in PRIMITIVE_TYPES:
+            raise ValueError(
+                "Expected schema of type `Record`, got instead primitive: {}".format(
+                    raw_input_schema
+                )
+            )
+
         record_format_identifier = b"\0\0\0\0"
-        schema_id = self._get_schema_id(cached_schema.fullname, str(cached_schema), **kwargs)
+        try:
+            schema_id = self._get_schema_id(
+                cached_schema.fullname, str(cached_schema), **kwargs
+            )
+        except AttributeError:
+            raise ValueError(
+                "Expected schema of type `Record`, got instead schema of type: {}".format(
+                    cached_schema.type
+                )
+            )
         data_bytes = self._avro_serializer.serialize(value, cached_schema)
```

Comment:
to be safe here, we could abstract the underlying avro serializer library

1. parse schema
  - _fast_avro_serializer.parse_schema, normal_avro_serializer.parse_schema
  - try:
      _fast_avro_serializer.parse_schema()
    exception Exception:
      raise SchemaParseException()
2. serialize
   try:
       _fast_avro_serializer.serialize()
   except Exception:
       SerializationException()
3. deserialize
    DeserializationException

Reply:
we could make a matrix for comparison:
                               fastavro            avro
missing property: raising Keyerror raise SchemaParseException
wrong format: ...

but I'm afraid it would take much time to traverse all the cases (especially comers cases which are difficult to think about, and may not be worth the efforts)